File: cpp/155.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
#include <sstream>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class MinStack {
  vector<int> stack;
  vector<int> mins;
public:
  void push(int x) {
    if(mins.empty() || x <= mins.back())
      mins.push_back(x);
    stack.push_back(x);
  }

  void pop() {
    if(stack.back() == mins.back())
      mins.pop_back();
    stack.pop_back();
  }

  int top() {
    return stack.back();
  }

  int getMin() {
    return mins.back();
  }
};

int main(){
  // auto sol = Solution();
}

File: python/155.py
class MinStack(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.mins = []
        self.stack = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        self.stack.append(x)
        if not self.mins or x <= self.mins[-1]:
            self.mins.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        v = self.stack.pop()
        if self.mins and self.mins[-1] == v:
            self.mins.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.mins[-1]
